### PR TITLE
Use `assert`s instead of `cast`s where possible

### PR DESCRIPTION
**Pull request by AlexWaygood**
There are many places in mypy's code where `cast`s are currently used unnecessarily. These can be replaced with `assert`s, which are much more type-safe, and more mypyc-friendly.

### CODE REVIEW

File: mypy/checkmember.py [hunk 1]
```diff
@@ -1150,7 +1151,8 @@ class B(A[str]): pass
             t = freshen_all_functions_type_vars(t)
             t = bind_self(t, original_type, is_classmethod=True)
             assert isuper is not None
-            t = cast(CallableType, expand_type_by_instance(t, isuper))
+            t = expand_type_by_instance(t, isuper)
```
Comment (JelleZijlstra) [inline mypy/checkmember.py:1154]: Wonder if we could enhance the type of `expand_type_by_instance` and `expand_type` to make these asserts unnecessary. Seems like it's close to `def expand_type[T: Type](t: T) -> T:`, but maybe it's not quite there.

File: mypy/semanal_enum.py [hunk 2]
```diff
@@ -108,7 +108,8 @@ class A(enum.Enum):
             # Error. Construct dummy return value.
             info = self.build_enum_call_typeinfo(var_name, [], fullname, node.line)
         else:
-            name = cast(StrExpr, call.args[0]).value
+            assert isinstance(call.args[0], StrExpr)
```
Comment (JelleZijlstra) [inline mypy/semanal_enum.py:111]: Might be cleaner to have `parse_enum_call_args` return this as another tuple member.

File: mypy/checkmember.py (hunk 1 above)
Comment (AlexWaygood) [reply mypy/checkmember.py:1154]: That looks plausible, but I don't really feel confident enough I fully understand that function to make changes there tbh.

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Moving this into draft: looking at this again, I think you're right; we can fairly easily improve the signature of `expand_type`, and by doing so remove the need for many of these `assert`s.

**Comment by AlexWaygood** [comment]
> Moving this back into draft: looking at this again, I think you're right; we can fairly easily improve the signature of `expand_type`, and by doing so remove the need for many of these `assert`s.

Here's a better solution for the `expand_type`/`expand_type_by_instance` ones:
- #14879

**Pull request merged**